src/goesrecv/hackrf_source.cc

```diff
@@ -0,0 +1,151 @@
+#include "hackrf_source.h"
+
+#include <pthread.h>
+
+#include <cstring>
+#include <iostream>
+
+#include <util/error.h>
+
+std::unique_ptr<HackRf> HackRf::open(uint32_t index) {
+  struct hackrf_device* dev = nullptr;
+  auto rv_init = hackrf_init();
+  if (rv_init < 0) {
+    std::cerr << "Unable to init hackrf" << std::endl;
+    exit(1);
+  }
+
+  auto rv = hackrf_open(&dev);
+  std::cout << "Is open?" << std::endl;
+
+  if (rv < 0) {
+    std::cerr << "Unable to open Hackrf device: "
+              << hackrf_error_name((enum hackrf_error)rv) << std::endl;
+    exit(1);
+  }
+
+  return std::make_unique<HackRf>(dev);
+}
+
+HackRf::HackRf(struct hackrf_device* dev) : dev_(dev) {
+  // Load list of supported sample rates
+  sampleRates_ = loadSampleRates();
+}
+
+HackRf::~HackRf() {
+  if (dev_ != nullptr) {
+    hackrf_close(dev_);
+    hackrf_exit();
+  }
+}
+
+std::vector<uint32_t> HackRf::loadSampleRates() {
+  return {8000000, 10000000, 12500000, 16000000, 20000000};
+}
+
+void HackRf::setFrequency(uint32_t freq) {
+  ASSERT(dev_ != nullptr);
+  auto rv = hackrf_set_freq(dev_, freq);
+  ASSERT(rv >= 0);
+}
+
+void HackRf::setSampleRate(uint32_t rate) {
+  ASSERT(dev_ != nullptr);
+  auto rv = hackrf_set_sample_rate(dev_, rate);
+  ASSERT(rv >= 0);
+  sampleRate_ = rate;
+}
+
+uint32_t HackRf::getSampleRate() const { return sampleRate_; }
+
+void HackRf::setIfGain(int gain) {
+  ASSERT(dev_ != nullptr);
+  auto rv = hackrf_set_lna_gain(dev_, gain);
+  ASSERT(rv >= 0);
+}
+
+void HackRf::setRfAmplifier(bool on) {
+  ASSERT(dev_ != nullptr);
+  auto rv = hackrf_set_amp_enable(dev_, on);
+  ASSERT(rv >= 0);
+}
+
+void HackRf::setBbGain(int gain) {
+  ASSERT(dev_ != nullptr);
+  auto rv = hackrf_set_vga_gain(dev_, gain);
+  ASSERT(rv >= 0);
+
+  auto rv_amp = hackrf_set_amp_enable(dev_, 1);
+  ASSERT(rv_amp >= 0);
```

Comment:
This conflicts with the `rf_amplifier` option. Should it be enabled by default with no way to turn it off?

Reply:
Good catch. Done.